Raise exception when RV32E instructions use x16-x31

eleanorLYJ opened the pull request:
Modify rv_decode() to enforce RV32E register constraints by validating rd, rs1, rs2, and rs3, ensuring they remain within x0-x15. If an instruction attempts to use x16-x31, trigger an illegal instruction exception, aligning with RISC-V privileged architecture behavior.

According to The RISC-V Instruction Set Manual Volume I: Unprivileged Architecture, Version 20240411:

"RV32E and RV64E ...  only registers x0-x15 are provided. All encodings specifying the other registers x16-x31 are reserved." 
 <div id='description'>
<h3>Summary by Bito</h3>
This pull request enhances the rv_decode() function by enforcing constraints on RV32E register usage, ensuring valid registers are limited to x0-x15. It improves instruction decoding robustness by triggering illegal instruction exceptions for invalid register usage. New variables and logic have been added for validation.
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
</div>

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: d470cd50185cc76be7c3946cb4f45d8eb3a70aa5 | Previous: 2889609096613a525deb4eea3ad8730c3a85e52b | Ratio |
|-|-|-|-|
| `Dhrystone` | `1316` Average DMIPS over 10 runs | `1300` Average DMIPS over 10 runs | `0.99` |
| `Coremark` | `918.562` Average iterations/sec over 10 runs | `929.824` Average iterations/sec over 10 runs | `1.01` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv reviewed (changes requested):
Don't touch `src/softfloat`.

eleanorLYJ committed:
Raise exception when RV32E instructions use x16-x31

Modify rv_decode() to enforce RV32E register constraints by validating
rd, rs1, rs2, ensuring they remain within x0-x15. If an instruction
attempts to use x16-x31, trigger an illegal instruction exception,
aligning with RISC-V privileged architecture behavior.

According to The RISC-V Instruction Set Manual Volume I: Unprivileged
Architecture, Version 20240411:

"RV32E and RV64E ...  only registers x0-x15 are provided. All encodings
specifying the other registers x16-x31 are reserved."

jserv commented on src/decode.c:
```diff
@@ -2026,7 +2026,16 @@ bool rv_decode(rv_insn_t *ir, uint32_t insn)
+        if (!op(ir, insn))
+            return false;
+
+        if (ir->rd > 15 || ir->rs1 > 15 || ir->rs2 > 15 || ir->rs3 > 15)
```
Add `unlikely` and comments.

visitorckw commented:
How about rewriting it this way to avoid repeating the same logic twice?

```diff
diff --git a/src/decode.c b/src/decode.c
index f65556d..a8ba70e 100644
--- a/src/decode.c
+++ b/src/decode.c
@@ -1983,6 +1983,8 @@ typedef bool (*decode_t)(rv_insn_t *ir, uint32_t insn);
 /* decode RISC-V instruction */
 bool rv_decode(rv_insn_t *ir, uint32_t insn)
 {
+    bool ret;
+
     assert(ir);
 
 #define OP_UNIMP op_unimp
@@ -2026,7 +2028,8 @@ bool rv_decode(rv_insn_t *ir, uint32_t insn)
         /* decode instruction (compressed instructions) */
         const decode_t op = rvc_jump_table[c_index];
         assert(op);
-        return op(ir, insn);
+        ret = op(ir, insn);
+        goto final;
     }
 #endif
 
@@ -2036,8 +2039,17 @@ bool rv_decode(rv_insn_t *ir, uint32_t insn)
     /* decode instruction */
     const decode_t op = rv_jump_table[index];
     assert(op);
-    return op(ir, insn);
+    ret = op(ir, insn);
 
 #undef OP_UNIMP
 #undef OP
+
+final:
+
+#if RV32_HAS(RV32E)
+    if (unlikely(ir->rd > 15 || ir->rs1 > 15 || ir->rs2 > 15 || ir->rs3 > 15))
+        return false;
+#endif
+
+    return ret;
 }

```

ChinYikMing commented:
The system emulation CI can be rerun and should be fixed. 

Related: #579

ChinYikMing commented:
> The system emulation CI can be rerun and should be fixed.
> 
> Related: #579

@jserv  Please merge #579 first, then rebase and the CI for this PR will be passed.

jserv reviewed (changes requested):
Rebase the latest `master` branch as @ChinYikMing requested.

jserv commented:
Can you check the failure of RISC-V Architectural Tests reported by CI?

jserv commented:
Thank @eleanorLYJ for contributing!
